feat: merge media controls.

csponge opened the pull request:
Initial commit to merge audio/video files. There are still a few bugs around widget sizing that need fixing.

CyanVoxel commented:
Thank you for your work on this so far! It should be fairly straightfoward to get the size of an image, but it differs a bit depending on what type of variable it is. [PIL images](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image) have `.width` and `.height` properties while [QPixmaps](https://doc.qt.io/qt-6/qpixmap.html) have `.width()` and `.height()` methods along with a `.size()` method which returns both as a QSize object. Depending on what you're doing it also may or may not be necessary to take into account the screen pixel ratio. Qt *usually* takes care of this for you, but images rendered from PIL do not.

As of your last commit I don't *think* I see the issue you're describing, but it might be part of the screenshot I've attached below. Since you're getting closer to finishing I also went ahead and jotted down some quick feedback based on the current state of the PR:

- The volume and playback sliders can't be clicked anywhere along the bar to set their values and instead can only be changed by dragging the handle. If it was just the volume silder then maybe this could pass, but it makes playback seeking rather difficult. (Maybe [this](https://python-forum.io/thread-10564.html) post could be of help?)
- Given certain video sizes the volume slider covers the playback timers and potentially gets cut off of the screen ![image](https://github.com/user-attachments/assets/195e0e51-3a6e-48e7-a3fa-a2e0056235e2)
- The toggle mute setting is not respected
- The autoplay setting is not respected
- The right-click context menu is missing
- Clicking on the preview outside of the controls opens the file - while this is the default behavior for other file types such as images, for media playback this is undesirable and the more intuitive behavior would be for the pause state to be toggled

csponge commented:
Hi,

Thanks for the comment! I'll be sure to work on those features/fixes.

On my original issue, here is a screenshot of the issue I'm having:
![overlay_oversized](https://github.com/user-attachments/assets/fb4cc12b-ee6f-4896-b0d0-4fa3d716e0c3)

In this case, the thumbnail image stops growing, while the rest of the media player widget keeps growing. This only happens when I make the preview panel very large.

Let me know if you have any questions.

CyanVoxel commented:
> In this case, the thumbnail image stops growing, while the rest of the media player widget keeps growing. This only happens when I make the preview panel very large.

Ahh, I see! It looks like this is due to the max resolution used for the preview panel thumb which is 512x512 in most places. It's unfortunately a messy hardcoded limit sprinkled around the codebase. I won't stop you if you're interested in improving it, but I also wouldn't burden you with that for this feature.

csponge commented:
Hi,

I just pushed a fix that should fix the sliders, autoplay, context menu, and the behavior when clicking on the media widget. I am running into a mypy issue that I'm not sure how to resolve yet.

On the thumbnail sizing, I guess I could also lock it to 512x512 for now. I'll try to get that in the next commit.

I have some questions on the your other comments:
- What should the behavior be when there is not enough space to display all the widgets? Do I need to resize them? Hide some of them? 
- How should toggle mute behave? I assume it would be similar to the autoplay setting. Start the media muted?

Thank you!

CyanVoxel commented:
Thank you for implementing the feedback so far!

> I am running into a mypy issue that I'm not sure how to resolve yet.

I see what you mean:
`opt.subControls = QStyle.SubControl.SC_SliderGroove | QStyle.SubControl.SC_SliderHandle`
`"QStyleOptionSlider" has no attribute "subControls" [attr-defined]`

From what I can tell, PySide6 is not implementing the `subControls` attribute like it says so in the [Qt documentation](https://doc.qt.io/qt-6/qstyleoptionslider-members.html). However from poking around in the code I also don't know if it would accept any `SubControl` value other than the ones beginning with `SO_`. In any case, removing this line doesn't appear to change any behavior?

> On the thumbnail sizing, I guess I could also lock it to 512x512 for now. I'll try to get that in the next commit.

Sounds good to me!

> What should the behavior be when there is not enough space to display all the widgets? Do I need to resize them? Hide some of them?

Would this be in cases like this? If so, I think hiding the volume slider when it would normally overlap with the timecode would be an alright solution in this case
![image](https://github.com/user-attachments/assets/7aae7120-a114-4d1e-89ed-953dc56c1f7f)

> How should toggle mute behave? I assume it would be similar to the autoplay setting. Start the media muted?

It should work the same way it currently does on `main` with videos, which in this case yes would be to start the media muted

***

I appreciate you implementing the feedback for jumping the seek position when clicking! However, I would expect it to move the handle to the clicked position as well, allowing you to continue holding down and dragging to seek. This is the behavior of any video/media player I can think of off the top of my head (e.g. YouTube, QuickTime, VLC, etc.)

CyanVoxel commented:
I just happened to notice the mypy errors here - `Translations.translate_qobject()` has been removed in favor of just using strings returned from `Translations["key_name"]` (for keys without placeholders).

So in the case of something like
```py
open_file_action = QAction(self)
Translations.translate_qobject(open_file_action, "file.open_file")
 ```
 this now becomes:
 ```py
open_file_action = QAction(Translations["file.open_file"])
 ```
 
 (I tried to comment on the code directly but GitHub didn't let me for some reason)

csponge commented:
Thanks for letting me know, I'll make sure to update that when pulling in the main branch.

I don't think I'm able to reproduce the issue you had when clicking the slider. Clicking the slider causes the handle to move to the position I clicked. Does the most recent commit fix this issue for you?

This might be a weird request, but would you be able to share a small video with me so I can test the widget overlap issues? If not, is there a resource I could find a sample video?

SkeleyM commented:
Hi, i know this isnt open for review, i just thought i'd mention i tried dynamically resizing the volume bar. it fixes the overlap issue but its pretty hard to control at smaller sizes.

![image](https://github.com/user-attachments/assets/5493ee55-bfb0-48b0-8087-b0ac15b2a8e2)
![image](https://github.com/user-attachments/assets/7d2b3d5d-e685-4b71-ba6d-ccbfd867b267)

heres the patch if youre interested
[dynamic.patch](https://github.com/user-attachments/files/19107490/dynamic.patch)

perhaps a better approach could be to make the volume slider vertical and popup when the media has a small width. however this does somewhat work

also here is a link to find stock videos of that size, just right click and press "Save Video As"
https://www.shutterstock.com/video/search?aspect_ratio=9%3A16

csponge commented:
Hey @SkeleyM,

Thanks for the patch! I tried it and it seemed to make it better. And I agree that it can get too small. Feel free to add more if you like :)

I'd be interested in trying a vertical slider. I'll see if I can get something like that working.

csponge commented:
Hello,

I refactored quite a bit in this last fix. In addition to the refactor, I added the ability to switch the orientation of the volume slider based on the size of the media player. Thank you @SkeleyM for the suggestion. How does it look?

Also, I made the volume slider only visible when hovering over the mute button. I thought that was a neat feature :)

There is still the issue on matching the size of the thumbnail for audio files. Need to continue researching there.

CyanVoxel commented:
I apologize for the inconvenience - a large refactor was recently merged (#844) that has moved several files throughout the codebase. This was done in order to correct some deep issues with the project's layout and was performed now as it seemed to be the best opportunity (as of 3/6/25) to disrupt as few open PRs as possible.

Unfortunately this PR is one that's been disrupted by this change. A rebase targeting `main` will be required to resolve the merge conflicts generated here. A merge commit is possible, however I *do not* recommend it as a rebase will be much smoother and much less work.
***
I also apologize for not getting back to you about the slider issue. While clicking it *does* move the handle to the pointer location, it does not let you continue dragging the handle from there. This makes it a bit awkward to control vs other media players.

Please let me know if there's anything else I can do and if you'd like any help with this. You've done some fantastic work so far and the last thing I want is to scare you off with a rebase or too many requests!

csponge commented:
Hi @CyanVoxel,

I just did a rebase and tested locally. I think I did it correctly... please let me know if I caused any issues.

As for the slider issue, I see what you mean now. I'll work on that next!

The only other thing I need help with is the sizing for thumbnail. If you have any good ways to fix it, please let me know. I will continue researching as well.

Thanks!

csponge commented:
Hi @xarvex,

Sorry, I hope I didn't create a bunch of work for you all. I have nothing to push at the moment, so I think you should be good to go.

xarvex commented:
To preserve the current state of this PR, I pushed to a new branch in our repo: https://github.com/TagStudioDev/TagStudio/tree/pull/805

I had to rebase from the point of the commit before the refactor (https://github.com/TagStudioDev/TagStudio/pull/805/commits/239a1b457e385256861f507e01cf4b75c8609c1a) and manually redo the commit that came after (https://github.com/TagStudioDev/TagStudio/pull/805/commits/098840c46633dce03f9286c7f7b986db6971e66e).

If all looks good there, I will force push to here.

CC @csponge @CyanVoxel

csponge commented:
Hi @xarvex,

I looked at the code and tested locally. Everything seems to be in place.

Thank you so much for fixing this! Feel free to push it.

CyanVoxel commented on src/tagstudio/qt/helpers/qslider_wrapper.py:
```diff
@@ -0,0 +1,41 @@
+# Copyright (C) 2024 Travis Abendshien (CyanVoxel).
```
```suggestion
# Copyright (C) 2025
```

CyanVoxel reviewed (commented):
I know this is still a draft, but I wanted to get a jump on checking out the underlying code now that the player itself is coming further along.

Overall I think you've done a great job with this feature, and the media player is currently much more usable than the previous video and audio player setup. There's just some comments I have on the current code structure, a few bugs, and some miscellaneous suggestions.

So far I have some concerns about the new media player code being somewhat "grafted" onto the existing code for the legacy video player and previous audio player code rather than outright replacing it. When I previously refactored and split up the preview panel, I made sure to refer to the video player code as "legacy" to denote that it's going to be replaced by a future combined media player and I named the audio player as "media" to serve as a jumping off point. I understand that this wasn't really communicated on my part though, so I get if there was some hesitation on replacing this code. But all the legacy video player code, including the `video_player.py` file, will no longer be required thanks to this new combined player and can/should be removed. Really after this point should just be code for "images", "animated images/gifs", and "media (audio/video)". 

In addition to this, I feel that the new `QStackedLayout` is not being fully taken advantage of and is currently just adding extra code with no current benefit. I would suggest looking into whether or not this layout can be used to replace the previous state management system that involved hiding widgets not in use with `switch_preview()` or to revert to the previous layout layout pattern. Currently there's one foot in the old system and one foot in this potential new system, when really it just needs to be one or the other.

Finally there's some bugs introduced here, with the current ones I've found being:
- The image and GIF previews no longer have right-click menus and cannot be opened via left click anymore
- The media player is missing the "Send to Trash" right-click option
- Videos do not loop, as was the previous default behavior
- The translation key for "Open file" is incorrect (see suggested change)

I hope this isn't too much, and please let me know if you have any questions or would like some help with this. You've done some great work so far and I would love to see  this feature make it to the end!

csponge commented:
Hi @CyanVoxel,

Sorry for not responding sooner, I was away last weekend. 

Thank you for doing a review! I appreciate all the feedback. I'll take some time today and work through your suggestions.

csponge commented on src/tagstudio/qt/widgets/preview/preview_thumb.py:
```diff
@@ -290,15 +303,15 @@ def _update_video_legacy(self, filepath: Path) -> dict:
             stats["height"] = image.height
             if success:
-                self.preview_vid.play(filepath_, QSize(image.width, image.height))
+                self.media_player.show()
```
Working on this, and I had an idea. The `QMediaPlayer` class has its own [metaData](https://doc.qt.io/qtforpython-6/PySide6/QtMultimedia/QMediaPlayer.html#id14) property. There is a lot of useful information _potentially_ available in the metadata, including video resolution. This could replace the use of `cv2` to get these stats, making the `_update_media` method a bit cleaner. There is a downside though.

From what I can tell, certain metadata isn't immediately available. We need to wait for the media to load before we can access it. This would require us to connect to a signal for metadata changes. Also, this would require us to change how we get file metadata in general. Instead of an immediate response, from `update_preview`, we would need to have a general signal to send the `stats` dictionary.

What are thoughts on this? It might be a bit overkill, so I'd also be fine leaving the `cv2` logic in place.

csponge commented:
Hi @CyanVoxel,

> Videos do not loop, as was the previous default behavior

Would it be beneficial to have looping as a setting, like autoplay? This is pretty easy to implement. Only downside is the setting needs to be translated.

CyanVoxel commented:
> Hi @CyanVoxel,
> 
> > Videos do not loop, as was the previous default behavior
> 
> Would it be beneficial to have looping as a setting, like autoplay? This is pretty easy to implement. Only downside is the setting needs to be translated.

This would be great! I would do it exactly like how the autoplay setting is done, and have the default behavior be set to true. For the translation key, I would make it:
```json
"media_player.loop": "Loop",
```
And remember that the en.json file is sorted alphabetically, so in this case the key will be immediately below the autoplay key.

csponge commented:
Tried adding the delete action with this last commit, but I'm not sure I got it working.

The loop setting works okay, but I can't get the delete action to show for some reason. I've added the action to both the `preview_gif` and the `media_player`, but it is not showing for either. Any ideas on this?

![image](https://github.com/user-attachments/assets/6060562e-5b34-47fd-baa0-5a99a208148f)

CyanVoxel commented:
> Tried adding the delete action with this last commit, but I'm not sure I got it working.
> 
> The loop setting works okay, but I can't get the delete action to show for some reason. I've added the action to both the `preview_gif` and the `media_player`, but it is not showing for either. Any ideas on this?

The delete action seems to appear and work just fine for me (macOS) 🤔 
Video:
<img width="204" alt="image" src="https://github.com/user-attachments/assets/6104e7d0-f2c5-47b2-a62c-1e2a4d338a5f" />
GIF:
<img width="185" alt="image" src="https://github.com/user-attachments/assets/3509c788-927f-469c-bb30-a32393cbac78" />

Although I've got a new issue where the program will launch with the preview panel now taking up half the width of the application window:
![Screenshot 2025-03-20 at 01 28 39](https://github.com/user-attachments/assets/8ba600ef-03b3-4888-b707-241a9a138d41)

csponge commented:
Okay, I think I know where I went wrong.

I was accidentally running an older version of the code somehow. This most recent commit should fix the issues.

Thanks!

csponge commented:
With my latest commit, I believe I've fixed one of my biggest issues with this feature. I think it should be good enough to be reviewed now.

Please let me know if there are any more issues.

Thanks!

CyanVoxel commented:
Everything seems to look good now! There's a couple of merge conflicts that have popped up due to the merging of #859, however if you'd like I can fix this up for you by updating this PR from main and applying the settings changes on top (I've already got the local commits ready if you're okay with me pushing them!)

Thank you so much for all your work on this!

csponge commented:
Yep, sounds great!  Thanks for fixing the conflicts.
